### ISSUE TITLE
paste to current window not working on Windows

### ISSUE THREAD

**Issue by ntzb** [hluk/CopyQ#2556]
I'm on Windows 11, and "paste to current window" almost never works, not from the main window nor from the tray.
why almost never? it very very rarely pastes from main window, but never from tray (it would paste from main window 2-3 times, then would not work again, tried many times).
also, it seems there's a focus obtaining issue, since the last window is not focused (you can tell by the titlebar color)

**To Reproduce**
set in preferences both options for paste (from history (main window) or from tray).
either double click an item on the main window, or select an item from the tray.

**Expected behavior**
copyq should paste to the "current" (last) window

**Version, OS and Environment**
 - CopyQ version 7.1.0
 - Windows 11 22H2 22621.1848

**Additional context**
I saw https://github.com/hluk/CopyQ/issues/1395 and tried:
1. viewing the copyq log, there's nothing that stands out.
   sometimes this shows in the log:
   ```
   Warning [2023-11-26 22:20:50.965] <Server-24612>: [default] QtWarning: QSystemTrayIcon::setVisible: No Icon set
   Warning [2023-11-26 22:20:57.981] <Server-24612>: [qt.qpa.fonts] QtWarning: Unable to enumerate family ' "Font Awesome 6 Free 
   Solid (CopyQ)" '
   ```
2.  running `ClipboardOwnerDebug_1001` - no visible errors
3. shift+insert in the last window works

any help or debugging tips would be very much appreciated.

**Comment by lightnin4u** [comment]
same problem here. It's definitely a lost focus problem.

**Comment by ntzb** [comment]
@hluk anything else I can do to help debug?

**Comment by hluk** [comment]
> anything else I can do to help debug?

Unfortunately no. I think the problem is either that triggering global shortcut or focusing other window breaks the focus.

If global shortcuts themselves do not cause the issue, the workaround would be to use "Copy next item"/"Copy previous item" actions or [Cycle Items - Quick](https://github.com/hluk/copyq-commands/tree/master/Global#cycle-items---quick) (see [README.md](https://github.com/hluk/copyq-commands/blob/master/README.md) for details on how to add the command).

You can also try to "Focus last window" option in History configuration tab in Preferences.

**Comment by singularity2000** [comment]
Yes, I think CopyQ should open as a popup. But I can't find this in settings.

**Comment by Chealer** [comment]
Ticket #1013 reports that installing the portable flavor on an affected system solved. It does not specify which version was first used though.

@ntzb: Do you have any evidence that this is specific to Windows 11?

<details>
  <summary>License</summary>
This comment and any other from myself on this ticket is offered under the terms of CC0 1.0.
</details>

**Comment by ntzb** [comment]
> [@ntzb](https://github.com/ntzb): Do you have any evidence that this is specific to Windows 11?

no, only tried it on Win 11 (and macOS but that was a long time ago)

**Comment by Chealer** [comment]
Thanks @ntzb 
In that case, specifying the version in the body suffices, in particular given that Windows 10 appears to be affected too.
Please retitle to something like:
> “Paste to current window“ generally fails silently on Microsoft Windows

**Comment by Chealer** [comment]
@singularity2000 wrote:
> Yes, I think CopyQ should open as a popup. But I can't find this in settings.

The system tray menu is already a popup, arguably. But there is no setting to have it open like @Chrysanthemum536 suggests.
Ditto has a ticket which [discusses a bit how its equivalent window could be implemented](https://github.com/sabrogden/Ditto/issues/830#issuecomment-3688717843).

**Comment by Chealer** [comment]
@hluk wrote:
> You can also try to "Focus last window" option in History configuration tab in Preferences.

What is that option supposed to do?

That preference is enabled on both of my CopyQ 13.0 installs, yet both are affected. Both are fresh installs on Windows 11 25H2. Out of between 50 and 150 attempts, only 5 worked, on my main PC. After each of these 5 times, I immediately tried to reproduce by activating the same items again right after, with the same target windows/fields. I have not managed to get a second successful paste again so far. The clips which I managed to paste were mostly rich text, but one was an image.

**Comment by Chealer** [comment]
@richlux reports the same symptoms with CopyQ 13 in ticket #3383, but with a major discovery: when CopyQ’s main window is displayed using a global shortcut, pasting (generally) succeeds. My understanding is that CopyQ 13 does not come with a global shortcut for that, but that richlux defined a custom shortcut for “Show main window under mouse cursor”, which allowed to work around. In fact, on both systems about which I reported on 2026-01-01, **setting a custom shortcut for either “Show main window under mouse cursor” or “Show the tray menu” allowed to work around**. If others can confirm, that could provide a way to make CopyQ usable on affected systems.

**Comment by richlux** [comment]
@Chealer "Show main window under mouse cursor" seems to always paste the text correctly for me, but "Show the tray menu", sometimes doesn't work.  In fact, it doesn't work for adding this comment in GitHub, but "Show main window under mouse cursor" does. I'm not sure why there would be any difference, but for me there is.

Rich

**Comment by ntzb** [comment]
the workaround indeed works, I tested it a bit with the tray option (which I use most).
btw, I notice that sometimes it works without the workaruond, but only if I manually put the focus fast enough where I want it to paste

**Comment by Chealer** [comment]
Thank you @ntzb
> btw, I notice that sometimes it works without the workaruond, but only if I manually put the focus fast enough where I want it to paste

Wow. I notice the same with both of my laptops, with both the main window and the tray menu! With the tray menu, sometimes the wrong clip is pasted though.

Sorry to insist, but given this bug’s severity, it would be appreciated if you could retitle this ticket.

**Comment by Chealer** [comment]
I confirm this also affects Windows 10. The bug reproduces on the Windows 10 install I can access most easily exactly like it does on Windows 11, easily, and can be worked around the same ways (shortcut or quick click).

@ntzb: Considering how severe this bug is, can you retitle this ticket as requested?

**Comment by Chealer** [comment]
Can this ticket please be ["pinned"](https://docs.github.com/en/issues/tracking-your-work-with-issues/administering-issues/pinning-an-issue-to-your-repository)?

Thank you very much @ntzb. If you don't mind, perhaps the following title would be even clearer:
> “Paste to current window“ generally fails silently on Microsoft Windows

**Comment by Chrysanthemum536** [comment]
got this problem on Cinnamon as well (for example, when pasting to the start menu)

**Comment by Chealer** [comment]
@Chrysanthemum536: If you're talking about an occasional issue, it must be a different problem. This one happens almost all the time unless using keyboard shortcuts.
If it happens all the time, it must still be a different bug with similar symptoms, since this one is only known to affect Windows. If [On Linux, global shortcuts, pasting or clipboard monitoring does not work](https://copyq.readthedocs.io/en/latest/known-issues.html#on-linux-global-shortcuts-pasting-or-clipboard-monitoring-does-not-work) does not match, search for tickets which do not have the `Windows` label.

**Comment by Chrysanthemum536** [comment]
it is the exact same problem as what I said here: https://github.com/hluk/CopyQ/issues/2737. What causes it I do not know.

**Comment by Chealer** [comment]
@Chrysanthemum536: No, what that describes is a behavior/effect, and a couple problems that can contribute to.

Given what I know about CopyQ and [the strange "quick click" exception @ntzb described](https://github.com/hluk/CopyQ/issues/2556#issuecomment-3750724829), I guess the bug this ticket reports is the result of the combination of:
1. the heuristic used on Windows to palliate the lack of a proper copying API, [discussed in another ticket](https://github.com/hluk/CopyQ/issues/3444#issuecomment-4134398226)
2. the tray menu and main window both removing focus from the target, as explained in #3540
3. failure to restore focus to the target on Windows (unless invoked via a keyboard shortcut).

If I am guessing right, just one of fixing #1, fixing #3 or stopping #2 would suffice to fix this bug. So you have a point that this bug is related to what you reported, but I think this bug is more focused on the third point, since the first 2 have probably affected all platforms from the start. I am hoping that #3 is a Windows-specific regression more likely to be solved.

In any case, this ticket does not make any mention of disappearing fields. If your target disappears and the scenario/impact is a significant one not already identified in #3540, please mention it there.

**Comment by ntzb** [comment]
I think I may have found what's happening, at least in my case.

The window CopyQ remembers as the paste target doesn't seem to be the application you came from.
For me it was consistently a titleless system window, and CopyQ appears to have activated it and pasted there, apparently successfully, just into the wrong place.
That might explain why nothing shows up in the log, and possibly why the tray path never works while the main window sometimes does.

Two things looked suspicious to me:

`isAltTabWindow()` only rejects a handful of window classes, so it seems to accept shell surfaces, cloaked store application windows and tool windows that `EnumWindows()` reports as visible but that you can't actually switch to.
`getCurrentWindow()` also appears to return the foreground window it just rejected when the enumeration finds no replacement, since the callback only assigns a window if it finds a suitable one.
If that's right, one bad sample could stick permanently: raising it makes it the foreground window, so every later sample would find it again.

PR: #3707

A Windows build with the change, if anyone wants to try it: https://github.com/ntzb/CopyQ/actions/runs/35205189797
The `.zip` artifact is the portable build (signing in to GitHub is needed to download artifacts).
Run it as `copyq.exe --session pastetest` so it doesn't touch your normal configuration, enable both paste options, and try the tray menu with the mouse.

It fixes it for me on Windows 11, but I don't know whether everyone here is hitting the same cause.
If it doesn't help you, the log now has a `Current window:` line that says which window was picked.
